File: my-folder/0124-binary-tree-maximum-path-sum/solution.cpp
class Solution {
    // int answer;
    int dfs(TreeNode*root,int&answer){
        if(!root){
            return 0;
        }
        int x = dfs(root->left,answer);
        int y = dfs(root->right,answer);
        answer = max(answer,x+y+root->val);
        return max(0,root->val+max(x,y));
    }
public:
    int maxPathSum(TreeNode* root) {
        int answer = INT_MIN;
        dfs(root,answer);
        return answer;
    }
};

File: my-folder/0124-binary-tree-maximum-path-sum/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    static int res;
    public int solve(TreeNode node){
        if(node==null){
            return 0;
        }
        else{
            int l=solve(node.left);
            int r=solve(node.right);
            res=Math.max(res,l+r+node.val);
            return Math.max(0,Math.max(l,r)+node.val);
        }

    }
    public int maxPathSum(TreeNode root) {
        res=root.val;
        solve(root);
        return res;
        
    }
}
